webhooks/virtualmachine/validation/scsi_controller.go

```diff
@@ -0,0 +1,259 @@
+	}
+
+	volumesPath := field.NewPath("spec", "volumes")
+
+	// Build controller map from spec.hardware.controllers
+	specControllers := make(map[int32]controllerInfo)
+	if vm.Spec.Hardware != nil {
+		for _, controller := range vm.Spec.Hardware.SCSIControllers {
+			specControllers[controller.BusNumber] = controllerInfo{
+				busNumber:   controller.BusNumber,
+				ctrlType:    controller.Type,
+				sharingMode: controller.SharingMode,
+				maxSlots:    getMaxSlotsForControllerType(controller.Type),
+				usedSlots:   0,
+			}
+		}
+	}
+
+	// Build device count map from status.hardware.controllers
+	statusDeviceCounts := make(map[int32]int32)
+	if vm.Status.Hardware != nil {
+		for _, controller := range vm.Status.Hardware.Controllers {
+			if controller.Type == vmopv1.VirtualControllerTypeSCSI {
+				statusDeviceCounts[controller.BusNumber] = int32(len(controller.Devices)) //nolint:gosec // disable G115
+			}
+		}
+	}
+
+	// Track volume assignments to controllers
+	volumeAssignments := make(map[int32][]string) // busNumber -> volume names
+
+	// Process each volume
+	for i, vol := range vm.Spec.Volumes {
+		if vol.PersistentVolumeClaim == nil {
+			continue
+		}
+
+		pvc := vol.PersistentVolumeClaim
+		volPath := volumesPath.Index(i)
+
+		// Determine which controller this volume should use
+		var targetBusNumber int32
+		var targetController controllerInfo
+		var hasTarget bool
+
+		// Check if a specific controller is requested
+		if pvc.ControllerBusNumber != nil {
+			targetBusNumber = *pvc.ControllerBusNumber
+
+			// Validate bus number range
+			if targetBusNumber < 0 || targetBusNumber >= maxSCSIControllers {
+				allErrs = append(allErrs, field.Invalid(
+					volPath.Child("persistentVolumeClaim", "controllerBusNumber"),
+					targetBusNumber,
+					fmt.Sprintf(invalidControllerBusNumberRangeFmt, maxSCSIControllers-1)))
+				continue
+			}
+
+			targetController, hasTarget = specControllers[targetBusNumber]
+			if !hasTarget {
+				allErrs = append(allErrs, field.Invalid(
+					volPath.Child("persistentVolumeClaim", "controllerBusNumber"),
+					targetBusNumber,
+					fmt.Sprintf(invalidControllerBusNumberDoesNotExist, targetBusNumber)))
+				continue
+			}
+		} else {
+			// Find appropriate controller based on application type
+			switch pvc.ApplicationType {
+			case vmopv1.VolumeApplicationTypeOracleRAC:
+				targetController, hasTarget = findControllerWithSharing(
+					vmopv1.VirtualControllerSharingModeNone,
+					specControllers,
+				)
+				if !hasTarget {
+					continue
+				}
+
+			case vmopv1.VolumeApplicationTypeMicrosoftWSFC:
+				targetController, hasTarget = findControllerWithSharing(
+					vmopv1.VirtualControllerSharingModePhysical,
+					specControllers,
+				)
+				if !hasTarget {
+					continue
+				}
+
+			default:
+				// Use first available SCSI controller (bus 0)
+				targetController, hasTarget = specControllers[0]
+				if !hasTarget {
+					// No controller at bus 0 - mutation webhook will add one
```

Comment:
Mutating webhook will run before validation webhook. So here, to validate as much as we can, maybe we should assign it to the last bus instead of the first one?

Reply:
```
When VM is created
mutation1 -> validation -> VM reconcile adding controllers by modifying spec -> mutation2
```

Oh mutation1 won't do anything, I keep forgetting this